webrtc/browser-to-browser.md

```diff
@@ -0,0 +1,122 @@
+On non-browser libp2p nodes doing a hole punch with TCP or QUIC, the libp2p node discovers its public address via the [identify] protocol.
+One can not use the [identify] protocol on browser nodes to discover ones public IP and port given that the browser uses a new port for each connection.
+For example say that the local browser node establishes a WebRTC connection C1 via browser-to-server to a server node and runs the [identify] protocol.
+The returned observed public port P1 will most likely (depending on the NAT) be a different port than the port observed on another connection C2.
+The only browser supported mechanism to discover ones public IP and port for a given connection is the non-libp2p protocol STUN.
+This is why this specification depends on STUN, and thus the availability of one or more STUN servers for _A_ and _B_ to discovery their public addresses.
+
+There are various publicly available STUN servers.
+As an alternative one may operate dedicated STUN servers for a given libp2p network.
+Further specification of the usage of STUN is out of scope for this specifitcation.
+
+As an aside, note that _A_ and _B_ do not need to use the same STUN server in order to establish a direct WebRTC connection.
+
+## Signaling Protocol
+
+The protocol id is `/webrtc-direct`.
+Messages are sent prefixed with the message length in bytes, encoded as an unsigned variable length integer as defined by the [multiformats unsigned-varint spec][uvarint-spec].
+
+``` protobuf
+syntax = "proto2";
+
+message Message {
+    // Specifies type in `data` field.
+    enum Type {
+        // String of `RTCSessionDescription.sdp`
+        OFFER = 0;
+        // String of `RTCSessionDescription.sdp`
+        ANSWER = 1;
+        // String of `RTCIceCandidate.toJSON()`
+        CANDIDATE = 2;
+    }
+
+    optional Type type = 1;
+    optional string data = 2;
+}
+```
+
+## Open Questions
+
+- Do we need a mechanism for browsers to advertise support for WebRTC browser-to-browser?
+
+  Say that browser B supports WebRTC browser-to-browser.
+  B listens via a relay and advertises its relayed address.
+  A discovers B's relayed address.
+  At this point A does not know whether B is a browser and thus supports WebRTC browser-to-browser, or whether B is e.g. a laptop potentially supporting TCP and QUIC hole punching via DCUtR but not WebRTC browser-to-browser.
+  In the latter case, A can not establish a direct connection to B.
+
+  Potential solution would be for B to advertise some protocol after the `/p2p-circuit` within its Multiaddr, e.g. `/ip6/<RELAY_IP>/udp/4001/p2p/<RELAY_PEER_ID>/p2p-circuit/webrtc-direct/p2p/<B_PEER_ID>`.
+  As an alternative, A can discover B's support via the identify protocol on the relayed connection or by optimistically opening a stream using the signaling protocol.
+  Both of the latter options would on failure happen at the expense of a wasted relayed connection.
```

Comment:
It also increases latency. Relayed connections take a fair bit to be established, plus more round trips until identify is exchanged.

I don't see anything wrong with adveristing `/webrtc-direct` as part of the address. It is a kind of transport in the end.

Reply:
> It also increases latency. Relayed connections take a fair bit to be established, plus more round trips until identify is exchanged.

Added in 6406947.

> I don't see anything wrong with adveristing `/webrtc-direct` as part of the address. It is a kind of transport in the end.

Discussion moved to https://github.com/libp2p/specs/pull/497#discussion_r1083223169.

Resolving here.